File: web/src/algorithms/sorting/shell-sort.ts
import type { AlgorithmImplementation, AlgorithmStep } from '@/lib/types/algorithm';

export const shellSort: AlgorithmImplementation = {
  id: 'shell-sort',
  name: 'Shell Sort',
  category: 'sorting',
  timeComplexity: { best: 'O(n log n)', average: 'O(n^(3/2))', worst: 'O(n²)' },
  spaceComplexity: 'O(1)',
  pseudocode: [
    { line: 0, text: 'procedure shellSort(A)' },
    { line: 1, text: '  n ← length(A)' },
    { line: 2, text: '  gap ← floor(n / 2)' },
    { line: 3, text: '  while gap > 0 do' },
    { line: 4, text: '    for i ← gap to n-1 do' },
    { line: 5, text: '      temp ← A[i]' },
    { line: 6, text: '      j ← i' },
    { line: 7, text: '      while j ≥ gap and A[j - gap] > temp do' },
    { line: 8, text: '        A[j] ← A[j - gap]' },
    { line: 9, text: '        j ← j - gap' },
    { line: 10, text: '      A[j] ← temp' },
    { line: 11, text: '    gap ← floor(gap / 2)' },
  ],

  generateSteps(input: number[]): AlgorithmStep[] {
    const array = [...input];
    const steps: AlgorithmStep[] = [];
    const n = array.length;
    const sortedIndices: number[] = [];

    let gap = Math.floor(n / 2);

    // Show initial state
    steps.push({
      type: 'highlight',
      array: [...array],
      highlightedIndices: [],
      secondaryIndices: [],
      sortedIndices: [],
      pseudocodeLine: 0,
      description: `Starting Shell Sort with ${n} elements`,
    });

    while (gap > 0) {
      steps.push({
        type: 'pass-complete',
        array: [...array],
        highlightedIndices: [],
        secondaryIndices: [],
        sortedIndices: [...sortedIndices],
        pseudocodeLine: 2,
        description: `Gap = ${gap}`,
      });

      for (let i = gap; i < n; i++) {
        const temp = array[i];
        let j = i;

        // Show element being inserted
        steps.push({
          type: 'select',
          array: [...array],
          highlightedIndices: [i],
          secondaryIndices: [],
          sortedIndices: [...sortedIndices],
          pseudocodeLine: 5,
          description: `Pick element A[${i}] = ${temp} for gap-insertion (gap=${gap})`,
        });

        while (j >= gap && array[j - gap] > temp) {
          // Compare
          steps.push({
            type: 'compare',
            array: [...array],
            highlightedIndices: [j, j - gap],
            secondaryIndices: [],
            sortedIndices: [...sortedIndices],
            pseudocodeLine: 7,
            description: `Compare A[${j - gap}] = ${array[j - gap]} > ${temp}? Yes`,
          });

          // Shift
          array[j] = array[j - gap];
          steps.push({
            type: 'swap',
            array: [...array],
            highlightedIndices: [j, j - gap],
            secondaryIndices: [],
            sortedIndices: [...sortedIndices],
            pseudocodeLine: 8,
            description: `Shift A[${j - gap}] = ${array[j - gap]} to position ${j}`,
          });

          j -= gap;
        }

        if (j >= gap) {
          // Show comparison that ends the loop
          steps.push({
            type: 'compare',
            array: [...array],
            highlightedIndices: [j, j - gap],
            secondaryIndices: [],
            sortedIndices: [...sortedIndices],
            pseudocodeLine: 7,
            description: `Compare A[${j - gap}] = ${array[j - gap]} > ${temp}? No, stop shifting`,
          });
        }

        // Place temp at final position
        array[j] = temp;
        steps.push({
          type: 'insert',
          array: [...array],
          highlightedIndices: [j],
          secondaryIndices: [],
          sortedIndices: [...sortedIndices],
          pseudocodeLine: 10,
          description: `Place ${temp} at position ${j}`,
        });
      }

      gap = Math.floor(gap / 2);

      steps.push({
        type: 'pass-complete',
        array: [...array],
        highlightedIndices: [],
        secondaryIndices: [],
        sortedIndices: [...sortedIndices],
        pseudocodeLine: 11,
        description: gap > 0 ? `Pass complete. Next gap = ${gap}` : 'Final pass complete (gap was 1)',
      });
    }

    // Mark all as sorted
    for (let i = 0; i < n; i++) sortedIndices.push(i);
    steps.push({
      type: 'sorted',
      array: [...array],
      highlightedIndices: [],
      secondaryIndices: [],
      sortedIndices: [...sortedIndices],
      pseudocodeLine: 0,
      description: 'Shell Sort complete. Array is sorted.',
    });

    return steps;
  },
};

File: apple/AestheticAlgorithm/Algorithms/Sorting/ShellSort.swift
import Foundation

struct ShellSort: AlgorithmExecutable {
    let id = "shell-sort"
    let name = "Shell Sort"
    let category = "sorting"
    let timeComplexity = (best: "O(n log n)", average: "O(n^(4/3))", worst: "O(n\u{00B2})")
    let spaceComplexity = "O(1)"

    let pseudocode: [(line: Int, text: String)] = [
        (0, "procedure shellSort(A)"),
        (1, "  n ← length(A)"),
        (2, "  gap ← n / 2"),
        (3, "  while gap > 0 do"),
        (4, "    for i ← gap to n-1 do"),
        (5, "      temp ← A[i]"),
        (6, "      j ← i"),
        (7, "      while j >= gap and A[j - gap] > temp do"),
        (8, "        A[j] ← A[j - gap]"),
        (9, "        j ← j - gap"),
        (10, "      A[j] ← temp"),
        (11, "    gap ← gap / 2"),
        (12, "  return A")
    ]

    func generateSteps(from input: [Int]) -> [AlgorithmStep] {
        var array = input
        var steps: [AlgorithmStep] = []
        let n = array.count

        if n <= 1 {
            return steps
        }

        steps.append(AlgorithmStep(
            type: .highlight,
            array: array,
            highlightedIndices: [],
            secondaryIndices: [],
            sortedIndices: [],
            pseudocodeLine: 0,
            description: "Starting Shell Sort on array of \(n) elements"
        ))

        var gap = n / 2

        while gap > 0 {
            // Show current gap
            steps.append(AlgorithmStep(
                type: .highlight,
                array: array,
                highlightedIndices: [],
                secondaryIndices: [],
                sortedIndices: [],
                pseudocodeLine: 3,
                description: "Gap = \(gap): performing insertion sort on gap-separated sublists"
            ))

            for i in gap..<n {
                let temp = array[i]
                var j = i

                // Show the element being inserted
                steps.append(AlgorithmStep(
                    type: .select,
                    array: array,
                    highlightedIndices: [i],
                    secondaryIndices: [],
                    sortedIndices: [],
                    pseudocodeLine: 5,
                    description: "Gap \(gap): inserting element \(temp) at index \(i)"
                ))

                while j >= gap && array[j - gap] > temp {
                    // Compare step
                    steps.append(AlgorithmStep(
                        type: .compare,
                        array: array,
                        highlightedIndices: [j, j - gap],
                        secondaryIndices: [],
                        sortedIndices: [],
                        pseudocodeLine: 7,
                        description: "Comparing: A[\(j - gap)]=\(array[j - gap]) > \(temp)? Yes"
                    ))

                    // Shift element
                    array[j] = array[j - gap]

                    steps.append(AlgorithmStep(
                        type: .swap,
                        array: array,
                        highlightedIndices: [j, j - gap],
                        secondaryIndices: [],
                        sortedIndices: [],
                        pseudocodeLine: 8,
                        description: "Shifting \(array[j]) from index \(j - gap) to index \(j)"
                    ))

                    j -= gap
                }

                if j != i {
                    // Place temp in its correct position
                    array[j] = temp

                    steps.append(AlgorithmStep(
                        type: .insert,
                        array: array,
                        highlightedIndices: [j],
                        secondaryIndices: [],
                        sortedIndices: [],
                        pseudocodeLine: 10,
                        description: "Placed \(temp) at index \(j)"
                    ))
                } else if j >= gap {
                    // No shift needed, show compare that stopped the loop
                    steps.append(AlgorithmStep(
                        type: .compare,
                        array: array,
                        highlightedIndices: [j, j - gap],
                        secondaryIndices: [],
                        sortedIndices: [],
                        pseudocodeLine: 7,
                        description: "Comparing: A[\(j - gap)]=\(array[j - gap]) > \(temp)? No, element stays"
                    ))
                }
            }

            // Pass complete for this gap
            steps.append(AlgorithmStep(
                type: .passComplete,
                array: array,
                highlightedIndices: [],
                secondaryIndices: [],
                sortedIndices: [],
                pseudocodeLine: 11,
                description: "Gap \(gap) pass complete. Array: \(array)"
            ))

            gap /= 2
        }

        // Final sorted state
        steps.append(AlgorithmStep(
            type: .sorted,
            array: array,
            highlightedIndices: Array(0..<n),
            secondaryIndices: [],
            sortedIndices: Array(0..<n),
            pseudocodeLine: 12,
            description: "Shell Sort complete"
        ))

        return steps
    }
}
